jena-text/src/main/java/org/apache/jena/query/text/TextIndexES.java

```diff
@@ -0,0 +1,394 @@
+        Map<String, Node> result = new HashMap<>();
+
+        if(uri != null) {
+            response = client.prepareGet(indexName, docDef.getEntityField(), uri).get();
+            if(response != null && !response.isSourceEmpty()) {
+                String entityField = response.getId();
+                Node entity = NodeFactory.createURI(entityField) ;
+                result.put(docDef.getEntityField(), entity);
+                Map<String, Object> source = response.getSource();
+                for (String field: docDef.fields()) {
+                    Object fieldResponse = source.get(field);
+
+                    if(fieldResponse == null) {
+                        //We wont return it.
+                        continue;
+                    }
+                    else if(fieldResponse instanceof List<?>) {
+                        //We are storing the values of fields as a List always.
+                        //If there are values stored in the list, then we return the first value,
+                        // else we do not include the field in the returned Map of Field -> Node Mapping
+                        List<?> responseList = (List<?>)fieldResponse;
+                        if(responseList != null && responseList.size() > 0) {
+                            String fieldValue = (String)responseList.get(0);
+                            Node fieldNode = NodeFactoryExtra.createLiteralNode(fieldValue, null, null);
+                            result.put(field, fieldNode);
+                        }
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    @Override
+    public List<TextHit> query(Node property, String qs) {
+
+        return query(property, qs, MAX_RESULTS);
+    }
+
+    /**
+     * Query the ElasticSearch for the given Node, with the given query String and limit.
+     * @param property the node property to make a search for
+     * @param qs the query string
+     * @param limit limit on the number of records to return
+     * @return List of {@link TextHit}s containing the documents that have been found
+     */
+    @Override
+    public List<TextHit> query(Node property, String qs, int limit) {
+
+        qs = parse(qs);
+        LOGGER.debug("Querying ElasticSearch for QueryString: " + qs);
+        SearchResponse response = client.prepareSearch(indexName)
+                .setTypes(docDef.getEntityField())
+                .setQuery(QueryBuilders.queryStringQuery(qs))
+                .setFrom(0).setSize(limit)
+                .get();
+
+        List<TextHit> results = new ArrayList<>() ;
+        for (SearchHit hit : response.getHits()) {
+
+            Node literal;
+            String field = (property != null) ? docDef.getField(property) : docDef.getPrimaryField();
+            List<String> value = (List<String>)hit.getSource().get(field);
+            if(value != null) {
+                literal = NodeFactory.createLiteral(value.get(0));
```

Comment:
CASE 1: I believe you mean that the text query has the syntax `X3`. Otherwise, yes, that's correct.

CASE 2: Yes, correct.

CASE 3: Yes, correct.

I can see that this is beginning to look difficult to implement. You have stored the values as arrays in ES, which makes sense for the purpose of keeping the index synchronized with the actual triples, but that makes it difficult to know which of the values actually matched the original query. I don't think you should perform e.g. string matching operations in Java code to find out which of the values in the array actually contained the query string. For one thing, that would cause problems with specialized Analyzers in case support for them are added later. Also I think it is the responsibility of the text index to know which value was matched.

For now I think it is better to leave `literal` as `null` instead of assigning it a potentially incorrect value. That's also what the Lucene backend does when the `storeValues` setting has not been enabled and thus original values are not available in the index.